python/demo/demo_mixed-poisson.py

```diff
@@ -258,38 +222,50 @@
 # and assemble it into the matrix `P`:
 
 # +
-a_p = fem.form(
-    [
-        [ufl.inner(sigma, tau) * dx + ufl.inner(ufl.div(sigma), ufl.div(tau)) * dx, None],
-        [None, ufl.inner(u, v) * dx],
-    ],
-    dtype=dtype,
+a_p = ufl.extract_blocks(
+    ufl.inner(sigma, tau) * dx + ufl.inner(ufl.div(sigma), ufl.div(tau)) * dx + ufl.inner(u, v) * dx
 )
-P = fem.petsc.assemble_matrix(a_p, bcs=bcs, kind="nest")
-P.assemble()
+
 # -
 
-# We now create a PETSc Krylov solver and set the preconditioner (`P`)
-# and operator (`A`) matrices:
+# We create finite element functions that will hold the $\sigma$ and $u$
+# solutions:
 
 # +
-ksp = PETSc.KSP().create(msh.comm)
-ksp.setOperators(A, P)
-ksp.setMonitor(lambda ksp, its, rnorm: print(f"Iteration: {its}, residual: {rnorm}"))
-ksp.setType("gmres")
-ksp.setTolerances(rtol=1e-8)
-ksp.setGMRESRestart(100)
+sigma, u = fem.Function(V, dtype=dtype), fem.Function(W, dtype=dtype)
 # -
 
-# To apply different solvers/preconditioners to the blocks of `P`, we
-# set the preconditioner to be a PETSc
-# [`fieldsplit`](https://petsc.org/release/manual/ksp/#sec-block-matrices)
+# We now create a linear problem solver for the mixed problem.
+# As we will use different preconditions for the individual blocks of the saddle point problem,
+# we specify the matrix kind to be "nest", so that we can use # [`fieldsplit`](https://petsc.org/release/manual/ksp/#sec-block-matrices)
 # (block) type and set the 'splits' between the $\sigma$ and $u$ fields.
 
+
+# +
+
+problem = fem.petsc.LinearProblem(
+    a,
+    L,
+    u=[sigma, u],
+    P=a_p,
+    kind="nest",
+    bcs=bcs,
+    petsc_options={
+        "ksp_type": "gmres",
+        "pc_type": "fieldsplit",
+        "pc_fieldsplit_type": "additive",
+        "ksp_rtol": 1e-8,
+        "ksp_gmres_restart": 100,
+    },
+)
+# -
+
+
 # +
-ksp.getPC().setType("fieldsplit")
-ksp.getPC().setFieldSplitType(PETSc.PC.CompositeType.ADDITIVE)
-nested_IS = P.getNestISs()
+# We get the nest index sets for the two fields
+nested_IS = problem.A.getNestISs()
+ksp = problem.solver
+ksp.setMonitor(lambda ksp, its, rnorm: print(f"Iteration: {its}, residual: {rnorm}"))
```

Comment:
Consider formatting, flushing and outputting only on the root MPI rank.
```suggestion
ksp.setMonitor(lambda ksp, its, rnorm: PETSc.Sys.Print(f"Iteration: {its:>4d}, residual: {rnorm:.3e}"))
```

If you don't want to use PETSc then something like
```suggestion
def monitor(ksp: PETSc.KSP, its: int, rnorm: float):
    if msh.comm.rank == 0:
        print(f"Iteration: {its:>4d}, residual: {rnorm:.3e}", flush=True)
ksp.setMonitor(monitor)
```